tta-server/app/logic/Logic.scala

```diff
@@ -1,32 +1,28 @@
 package logic
 
-import models.Action
-import models.ActionId
-import models.Bronze
-import models.BuildBronze
-import models.DerivedGameState
-import models.DerivedPlayerState
-import models.GameState
-import models.PlayerIndex
-import models.PlayerState
+import models._
 
 import scala.collection.immutable
 
 object Logic {
   
   def deriveGameState(gameState: GameState): DerivedGameState = {
-    val defaultCandidateActions: immutable.Seq[Action] = List(BuildBronze)
+    val buildCandidateActionMap = gameState.activePlayerState.researchedBuildings.map (
+      building => {
+        ActionId("build" + building.prettyName) -> Building.generateBuildAction(building)
+      } //why does flatMap have an error here?
+    ).toMap
 
-    val allowedActions = defaultCandidateActions.filter { action =>
+    val increasePopulationCandidateActionMap = Map(ActionId("increasePopulation") -> Population.generateIncreasePopulationAction(gameState.activePlayerState))
+
+    val candidateActionMap = buildCandidateActionMap ++ increasePopulationCandidateActionMap
+
+    val allowedActions = candidateActionMap.filter { case (index, action)=>
       action.isValid(gameState)
```

Comment:
maybe not to be fixed in this revision, but I do not think that there should be an `isValid` function on actions

instead we should structure the action generation in such a way that only allowable actions get generated in the first place.

the reason is that our action generation is built in such a way that there are three types of actions:
1. those that never get generated (eg a build for an unresearched building)
2. those that get generated but filtered out (eg a build for a researched building that you don't have enough resources to build)
3. those that get generated and not filtered out

this means that filtering is really happening in two different places (steps 1 and 2), which makes it harder to reason about